dataframe.py

```diff
@@ -0,0 +1,270 @@
+        return (isinstance(other, Categorical) and
+                self.index_type == other.index_type and
+                self.category_type == other.category_type and
+                self.ordered == other.ordered)
+
+    def to_string(self):
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
+
+
+# ----------------------------------------------------------------------
+# DataFrame: the main public API
+
+
+class DataFrame(ABC, Mapping):
+    """
+    An abstract data frame base class.
+
+    A "data frame" represents an ordered collection of named columns. A
+    column's "name" is permitted to be any hashable Python value, but strings
+    are common. Names are not required to be unique. Columns may be accessed by
+    name (when the name is unique) or by position.
+    """
+
+    def __dataframe__(self):
+        """
+        Idempotence of data frame protocol
+        """
+        return self
+
+    def __iter__(self):
+        # TBD: Decide what iterating should return
+        return iter(self.column_names)
+
+    def __len__(self):
+        return self.num_rows
+
+    @property
+    @abstractmethod
+    def num_columns(self):
+        """
+        Return the number of columns in the DataFrame
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def num_rows(self):
+        """
+        Return the number of rows in the DataFrame (if known)
+        """
+        pass
+
+    @abstractmethod
+    def iter_column_names(self) -> Iterable[Any]:
+        """
+        Return the column names as an iterable
+        """
+        pass
+
+    # TODO: Should this be a method or property?
+    @property
+    @abstractmethod
+    def column_names(self) -> Sequence[Any]:
+        """
+        Return the column names as a materialized sequence
+        """
+        pass
+
+    # TODO: Should this be a method or property?
+    @property
+    def row_names(self) -> Sequence[Any]:
```

Comment:
I'm -1 on row names / indices. IMO that shouldn't be part of the standard (of course pandas and any other project can have this as a feature, but shouldn't be here).

Of course we could have this optional, but I think it's more a pandas thing, and we shouldn't add complexity to the standard.

Reply:
R's DataFrame also has row labels, so it's not specific to pandas. Are there any dataframe implementations that don't have row labels?

Reply:
The way I look at it if some implementations return `None` while others return a `Sequence` it is not very harmful.

As one can of worms, `row_names` may need to return an object with the same API as columns. So this might better return `Column`

Reply:
I think most dataframe implementations have row labels. But in the Python world I think that is because everybody is trying to replicate the pandas API for compatibility.

Having row labels is surely a good idea for many dataframe implementations. My point is that not having them is also a good idea to me. So, I would leave them out of this proposal, at least initially.

IIRC @maartenbreddels mentioned that he'd make Vaex behave more like SQL regarding indexing. If I had to implement a dataframe library from scratch I'd also have indices as regular columns that can be indexed. There would surely be some drawbacks compared to pandas, but I think the API for users would be much simpler, and I prefer that.

In any case, not trying to say that projects shouldn't have row labels. Just that my preference is to not include them here. It narrows the discussion, and it avoids biasing towards the R/pandas approach.

Reply:
Do we want any kind of invariant around `np.asarray(dataframe)`? To me the most natural case is something like `np.concat([col for col in df.values()], axis=1)`

But if you're storing your row labels "in band" (as part of the columns in the dataframe) then your row labels will be mixed with the values.

Reply:
Another note / leading question here: What's the argument for having `DataFrame.row_names` but not `Column.row_names`? I think if we want them for DataFrame then we also want them for Column (likewise if we don't want them).

Reply:
I'm also -1 on row names / indices. If it becomes an optional part of the protocol but say Pandas implements it, I imagine all of the other libraries will end up having their hands forced to support it from users wanting Pandas compatibility as opposed to building a new DataFrame standard.

Reply:
> as opposed to building a new DataFrame standard

The intention of this interface is for libraries to be able to expose the data that they contain to consumers.

Both R and pandas data frames have row names. So I don't see a big downside of having an API that allows them to make their row names available to consumers. It **does not mean** that other libraries need to implement row names

Reply:
> Both R and pandas data frames have row names. So I don't see a big downside of having an API that allows them to make their row names available to consumers. It **does not mean** that other libraries need to implement row names

While it isn't required from the protocol perspective, people will end up writing code using these optional features unnecessarily, i.e. in Pandas using `df.reset_index()[...]` versus `df.iloc[...]`. Instead of these features being used specifically when necessary, they end up getting used everywhere which becomes almost a forcing hand for libraries to have to support the feature.

Reply:
If a consumer writes code that assumes there are always row names, even though the interface says that they are optional, isn't that the consumer's problem? The abstract API is a contract for producer and consumer, so it wouldn't be appropriate for consumers to de facto violate the contract by making an optional feature required.

Reply:
It's the consumer's problem until there's enough people with the same problem (not wanting to rewrite code that assumes row names) that it becomes a problem for producers.

For what it's worth, I fought against having row indices in cuDF and ultimately lost the battle from the overwhelming amount of libraries we wanted to integrate with and user code assuming row indices purely because they were built against Pandas originally and it became too large of a battle to continue fighting.

I'm still against row indices, but would rather them be required than optional.

Reply:
Sorry, I guess that one's mostly my fault.

In R, row names are optional and it doesn't seem to cause problems. In pandas the `.index` has a functional purpose beyond purely communicating data, that may be part of why people are asking for it. In this interface, there is no computational functionality, purely one-way data transfer, so that seems less problematic.

Reply:
> In pandas the `.index` has a functional purpose beyond purely communicating data, that may be part of why people are asking for it.

In our case an Index is just a Column that we provide the syntactic sugar expected from an Index around it. If someone wants to grab rows by index value, we currently build a hash table and then probe that hash table each time (we have plans to memoize it in the future). No one has pointed this out or complained about performance of this thus far (outside of CUDA developers writing the libcudf C++ library 😆), so I don't think it has to do with computational functionality and is purely the data transfer aspects.